File: bot/exts/fun/wolfram.py
import datetime
from typing import Tuple, Union

from discord import Embed
from discord.ext import commands
from yarl import URL

from bot.utils.constants import Names, Tokens
from bot.utils.embed import ErrorEmbed

IMAGE_PARAMS = {
    "appid": Tokens.WOLFRAM_ID,
    "background": "2F3136",
    "foreground": "white",
    "layout": "labelbar",
    "fontsize": "23",
    "width": "700",
}

SHORT_PARAMS = {"appid": Tokens.WOLFRAM_ID}


class WolframCommands(commands.Cog):
    """
    Wolfram Category cog, containing commands related to the WolframAlpha API.

    Commands
        └ wolfram / wa
            ├ image         Fetch the response to a query in the form of an image.
            └ text          Fetch the response to a query in a short phrase.
    """

    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot
        self.image_url = "http://api.wolframalpha.com/v1/simple"
        self.short_url = "http://api.wolframalpha.com/v1/result"

    async def web_request(self, url: str, params: dict) -> Tuple[Union[URL, str, dict], int]:
        """Web request handler for wolfram group of commands."""
        async with self.bot.http_session.get(url=url, params=params) as resp:
            if resp.status == 200:
                try:
                    return await resp.text(), resp.status
                except Exception:
                    return resp.url, resp.status
            elif resp.status == 501:
                return "Sorry, the API could not understand your input", resp.status
            elif resp.status == 400:
                return "Sorry, the API did not find any input to interpret", resp.status

    @commands.group(name="wolfram", aliases=("wa",), invoke_without_command=True)
    async def wolfram_group(self, ctx: commands.Context) -> None:
        """Commands for wolfram."""
        await ctx.send_help(ctx.command)

    @wolfram_group.command(name="image", aliases=("i", "im"))
    async def wolfram_image(self, ctx: commands.Context, *, query: str) -> None:
        """Sends wolfram image corresponding to the given query."""
        await ctx.message.delete()
        IMAGE_PARAMS["i"] = query

        async with ctx.typing():
            response = await self.web_request(url=self.image_url, params=IMAGE_PARAMS)

            if isinstance(response[0], str):
                embed = ErrorEmbed(
                    description=response[0], author=ctx.author
                )
            else:
                embed = Embed(
                    title=f"Query: {query}",
                    timestamp=datetime.datetime.utcnow()
                ).set_footer(text=f"Invoked by {ctx.author.name}", icon_url=ctx.author.avatar_url)

                embed.set_image(url=response[0])
                embed.add_field(
                    name="Cannot see image?",
                    value=f"[Click here](https://www.wolframalpha.com/input/?i={query.replace(' ', '+')})",
                )
            await ctx.send(embed=embed, delete_after=60)

    @wolfram_group.command(name="short", aliases=("s", "sh"))
    async def wolfram_short(self, ctx: commands.Context, *, query: str) -> None:
        """Sends wolfram image corresponding to the given query."""
        await ctx.message.delete()
        SHORT_PARAMS["i"] = query

        async with ctx.typing():
            response = await self.web_request(url=self.short_url, params=SHORT_PARAMS)

            if response[1] in [400, 501]:
                embed = ErrorEmbed(
                    description=response[0], author=ctx.author
                )
            else:
                embed = Embed(
                    title=f"Query: {query}",
                    description=f"[**{response[0]}**](https://www.wolframalpha.com/input/?i={query.replace(' ', '+')})",
                    timestamp=datetime.datetime.utcnow()
                ).set_footer(text=f"Invoked by {ctx.author.name}", icon_url=ctx.author.avatar_url)

            await ctx.send(embed=embed, delete_after=60)


def setup(bot: commands.Bot) -> None:
    """Load the Wolfram cog."""
    bot.add_cog(WolframCommands(bot))


# Add function to hide the token

File: bin/interactions/fun/wolfram.dart
import 'package:dio/dio.dart';
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_interactions/interactions.dart';

import '../../obsidian_dart.dart' show botInteractions;
import '../../utils/constants.dart';

class FunWolframInteractions {
  final imageUrl =
      'http://api.wolframalpha.com/v1/simple?appid=${Tokens.WOLFRAM_ID}&layout=labelbar&background=0C0B0B&foreground=white&width=400';
  final shortUrl =
      'http://api.wolframalpha.com/v1/result?appid=${Tokens.WOLFRAM_ID}';
  late Map<String, String> shortParams;
  final _dio = Dio();

  FunWolframInteractions() {
    botInteractions.registerSlashCommand(SlashCommandBuilder(
      'wolfram',
      'Wolfram group of commands.',
      [
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'short',
          'Retrieves short answer for the question',
          options: [
            CommandOptionBuilder(CommandOptionType.string, 'query',
                'The query to ask to the Wolfram API.',
                required: true)
          ],
        )..registerHandler(wolframShortSlashCommand),
        CommandOptionBuilder(
          CommandOptionType.subCommand,
          'image',
          'Retrieves image for the question',
          options: [
            CommandOptionBuilder(CommandOptionType.string, 'query',
                'The query to ask to the Wolfram API.',
                required: true)
          ],
        )..registerHandler(wolframImageSlashCommand)
      ],
    ));
  }

  Future<String> webRequestHandler(
      String url, Map<String, String> params) async {
    var response = await _dio.get(url, queryParameters: params);
    if (response.statusCode == 200) {
      return response.data.toString();
    } else if (response.statusCode == 400) {
      return 'Sorry, the API did not find any input to interpret.';
    } else if (response.statusCode == 501) {
      return 'Sorry, the API could not understand your input.';
    } else {
      return 'Unidentified error! Status code: ${response.statusCode}, Error: ${response.data}';
    }
  }

  EmbedBuilder createWolframEmbed(
      SlashCommandInteractionEvent event, String title, String desc) {
    return EmbedBuilder()
      ..title = title
      ..description = desc
      ..color = DiscordColor.sapGreen
      ..addFooter((footer) {
        footer.text = 'Requested by ${event.interaction.userAuthor?.username}';
        footer.iconUrl = event.interaction.userAuthor?.avatarURL();
      });
  }

  Future<void> wolframShortSlashCommand(
      SlashCommandInteractionEvent event) async {
    await event.acknowledge();
    final query = event.getArg('query').value;
    shortParams = {'i': query};

    final response = await webRequestHandler(shortUrl, shortParams);
    final embed =
        createWolframEmbed(event, 'Query: $query', 'Response:\n **$response**');
    await event.respond(MessageBuilder.embed(embed));
  }

  Future<void> wolframImageSlashCommand(
      SlashCommandInteractionEvent event) async {
    await event.acknowledge();
    final query = event.getArg('query').value.toString();

    final response = '$imageUrl&i=${query.trim().replaceAll(' ', '+')}';
    final embed = createWolframEmbed(
        event, 'Query: $query', 'Open original to view higher res image.');
    embed.imageUrl = response;

    await event.respond(MessageBuilder.embed(embed));
  }
}
